File: cpp/34.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  vector<int> searchRange(int A[], int n, int target) {
    if(n == 0)
      return vector<int>{-1, -1};
    int l = distance(A, lower_bound(A, A+n, target));
    int r = distance(A, prev(upper_bound(A, A+n, target)));
    return A[l] == target ? vector<int>{l, r} : vector<int>{-1, -1};
  }
};


int main(){
  auto sol = Solution();
  // int a[] = {4,5,6,7,0,1,2};
  // int a[] = {0,1,2,4,5,6,7};
  // int a[] = {1, 3};
  // cout << sol.search(a, 2, 2);
}

File: python/34.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if not nums or target < nums[0]:
            return [-1, -1]

        n = len(nums)
        # nums[l] <= target < nums[r]
        l, r = 0, n
        while l+1 < r:
            m = (l+r) / 2
            if nums[m] <= target:
                l = m
            else:
                r = m

        if l == n or nums[l] != target:
            return [-1, -1]
        right = r = l
        if nums[0] == target:
            return [0, r]
        l = 0

        # nums[l] < target <= nums[r]
        while l+1 < r:
            m = (l+r) / 2
            if nums[m] < target:
                l = m
            else:
                r = m
        return [r, right]
